src/io/config.cpp

```diff
@@ -307,12 +307,21 @@ void Config::CheckParamConflict() {
   }
   // Check max_depth and num_leaves
   if (max_depth > 0) {
-    int full_num_leaves = static_cast<int>(std::pow(2, max_depth));
+    double full_num_leaves = std::pow(2, max_depth);
     if (full_num_leaves > num_leaves
         && num_leaves == kDefaultNumLeaves) {
       Log::Warning("Accuracy may be bad since you didn't set num_leaves and 2^max_depth > num_leaves");
     }
-    num_leaves = std::min(num_leaves, 2 << max_depth);
+    
+    const int MAX_LEAVES = 2 << 16;
```

Comment:
Sorry, why do we use `16` here?

Reply:
Only because @guolinke suggested it https://github.com/microsoft/LightGBM/issues/2357#issuecomment-530402138

Reply:
Ah, OK!
Why do we perform this check only in case of `if (max_depth > 0)`?

Reply:
No reason I guess, I put it there while I was adding the guard for the overflow. But now that you mention it, I don't see a reason why it couldn't be outside this block

Reply:
I think this check should be performed independently of the `max_depth` value. In this case, the check and `kMaxNumLeaves` can be moved to `config.h` and simplify the code.

Reply:
I moved the constant to config.h and the check outside the max_depth check.

I'm not sure where I'd place the check inside `config.h` If you want, feel free to add commits to this pr directly

Reply:
> I'm not sure where I'd place the check inside config.h If you want, feel free to add commits to this pr directly

Unfortunately, I have no permission to commit in your branch.
https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork

I would like to replace `if (num_leaves > kMaxNumLeaves)` in `config.cpp` with `// check = <=131072` in `config.h` so that the check will be created automatically and users will be aware of the max value early during reading the docs.

Reply:
updated